File: src/Slang.Sdk/Layer2-Binding/Reflection-API/TypeReflection.cs
using Slang.Sdk.Interop;
using static Slang.Sdk.Interop.StrongTypeInterop;
using static Slang.Sdk.Interop.SlangNativeInterop;
using static Slang.Sdk.Interop.StringMarshaling;

namespace Slang.Sdk.Binding
{
    internal unsafe class TypeReflection : Reflection
    {
        internal override Reflection? Parent { get; }
        internal override TypeReflectionHandle Handle { get; }

        internal TypeReflection(Reflection parent, TypeReflectionHandle handle)
        {
            Parent = parent;
            Handle = handle;
        }

        internal TypeKind GetKind()
        {
            string? error = null;
            return (TypeKind)Call(() => SlangNativeInterop.TypeReflection_GetKind(Handle, out error), () => error);
        }

        internal uint GetFieldCount()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_GetFieldCount(Handle, out error), () => error);
        }

        internal VariableReflection GetFieldByIndex(uint index)
        {
            string? error = null;
            return new VariableReflection(this, Call(() => StrongTypeInterop.TypeReflection_GetFieldByIndex(Handle, index, out error), () => error));
        }

        internal bool IsArray()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_IsArray(Handle, out error), () => error);
        }

        internal TypeReflection UnwrapArray()
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongTypeInterop.TypeReflection_UnwrapArray(Handle, out error), () => error));
        }

        internal nuint GetElementCount()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_GetElementCount(Handle, out error), () => error);
        }

        internal TypeReflection GetElementType()
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongTypeInterop.TypeReflection_GetElementType(Handle, out error), () => error));
        }

        internal uint GetRowCount()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_GetRowCount(Handle, out error), () => error);
        }

        internal uint GetColumnCount()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_GetColumnCount(Handle, out error), () => error);
        }

        internal ScalarType GetScalarType()
        {
            string? error = null;
            return (ScalarType)Call(() => SlangNativeInterop.TypeReflection_GetScalarType(Handle, out error), () => error);
        }

        internal TypeReflection GetResourceResultType()
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongTypeInterop.TypeReflection_GetResourceResultType(Handle, out error), () => error));
        }

        internal int GetResourceShape()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_GetResourceShape(Handle, out error), () => error);
        }

        internal int GetResourceAccess()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_GetResourceAccess(Handle, out error), () => error);
        }

        internal string GetName()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_GetName(Handle, out error), () => error);
        }

        internal uint GetUserAttributeCount()
        {
            string? error = null;
            return Call(() => SlangNativeInterop.TypeReflection_GetUserAttributeCount(Handle, out error), () => error);
        }

        internal AttributeReflection GetUserAttributeByIndex(uint index)
        {
            string? error = null;
            return new AttributeReflection(this, Call(() => StrongTypeInterop.TypeReflection_GetUserAttributeByIndex(Handle, index, out error), () => error));
        }

        internal AttributeReflection FindAttributeByName(string name)
        {
            string? error = null;
            return new AttributeReflection(
                this,
                Call(() => StrongTypeInterop.TypeReflection_FindAttributeByName(Handle, name, out error), () => error));
        }

        internal TypeReflection ApplySpecializations(GenericReflection genRef)
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongTypeInterop.TypeReflection_ApplySpecializations(Handle, genRef.Handle, out error), () => error));
        }

        internal GenericReflection GetGenericContainer()
        {
            string? error = null;
            return new GenericReflection(this, Call(() => StrongTypeInterop.TypeReflection_GetGenericContainer(Handle, out error), () => error));
        }
    }
}

File: src/Slang.Net.CPP/TypeReflection.cpp
// Define this before including any Windows headers to avoid conflicts
#define NOMINMAX
#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif

#include "TypeReflection.h"
#include "StringUtils.h"
#include "VariableReflection.h"
#include "Attribute.h"
#include "GenericReflection.h"
#include <msclr/marshal.h>

using namespace SlangNative;

namespace Slang::Cpp
{
    // Constructor
    TypeReflection::TypeReflection(void* native)
    {
        m_NativeTypeReflection = native;
        m_bOwnsNative = false; // We don't own the native pointer in this case
    }

    // Destructor
    TypeReflection::~TypeReflection()
    {
        this->!TypeReflection();
    }

    // Finalizer
    TypeReflection::!TypeReflection()
    {
		SlangNative::TypeReflection_Release(m_NativeTypeReflection);
        m_NativeTypeReflection = nullptr;
    }    
    
    // Properties
    TypeKind TypeReflection::Kind::get()
    {
        if (!m_NativeTypeReflection) return TypeKind::None;
        return static_cast<TypeKind>(TypeReflection_GetKind(m_NativeTypeReflection));
    }

    unsigned int TypeReflection::FieldCount::get()
    {
        if (!m_NativeTypeReflection) return 0;
        return TypeReflection_GetFieldCount(m_NativeTypeReflection);
    }

    VariableReflection^ TypeReflection::GetFieldByIndex(unsigned int index)
    {
        if (!m_NativeTypeReflection) return nullptr;
        void* field = TypeReflection_GetFieldByIndex(m_NativeTypeReflection, index);
        return field ? gcnew VariableReflection(field) : nullptr;
    }

    bool TypeReflection::IsArray::get()
    {
        if (!m_NativeTypeReflection) return false;
        return TypeReflection_IsArray(m_NativeTypeReflection);
    }

    TypeReflection^ TypeReflection::UnwrapArray()
    {
        if (!m_NativeTypeReflection) return nullptr;
        void* unwrapped = TypeReflection_UnwrapArray(m_NativeTypeReflection);
        return unwrapped ? gcnew TypeReflection(unwrapped) : nullptr;
    }    
    System::UIntPtr TypeReflection::ElementCount::get()
    {
        if (!m_NativeTypeReflection) return System::UIntPtr::Zero;
        return System::UIntPtr(TypeReflection_GetElementCount(m_NativeTypeReflection));
    }    
    System::UIntPtr TypeReflection::TotalArrayElementCount::get()
    {
        if (!m_NativeTypeReflection) return System::UIntPtr::Zero;
        // Note: This function may not be available in native interface, using ElementCount for now
        return System::UIntPtr(TypeReflection_GetElementCount(m_NativeTypeReflection));
    }
    TypeReflection^ TypeReflection::ElementType::get()
    {
        if (!m_NativeTypeReflection) return nullptr;
        void* elementType = SlangNative::TypeReflection_GetElementType(m_NativeTypeReflection);
        return elementType ? gcnew TypeReflection(elementType) : nullptr;
    }

    unsigned int TypeReflection::RowCount::get()
    {
        if (!m_NativeTypeReflection) return 0;
        return SlangNative::TypeReflection_GetRowCount(m_NativeTypeReflection);
    }

    unsigned int TypeReflection::ColumnCount::get()
    {
        if (!m_NativeTypeReflection) return 0;
        return SlangNative::TypeReflection_GetColumnCount(m_NativeTypeReflection);
    }

    Slang::Cpp::ScalarType TypeReflection::ScalarType::get()
    {
        if (!m_NativeTypeReflection) return Slang::Cpp::ScalarType::None;
        return static_cast<Slang::Cpp::ScalarType>(SlangNative::TypeReflection_GetScalarType(m_NativeTypeReflection));
    }    
    TypeReflection^ TypeReflection::ResourceResultType::get()
    {
        if (!m_NativeTypeReflection) return nullptr;
        void* resultType = SlangNative::TypeReflection_GetResourceResultType(m_NativeTypeReflection);
        return resultType ? gcnew TypeReflection(resultType) : nullptr;
    }

    Slang::Cpp::ResourceShape TypeReflection::ResourceShape::get()
    {
        if (!m_NativeTypeReflection) return Slang::Cpp::ResourceShape::None;
        return static_cast<Slang::Cpp::ResourceShape>(SlangNative::TypeReflection_GetResourceShape(m_NativeTypeReflection));
    }

    Slang::Cpp::ResourceAccess TypeReflection::ResourceAccess::get()
    {
        if (!m_NativeTypeReflection) return Slang::Cpp::ResourceAccess::None;
        return static_cast<Slang::Cpp::ResourceAccess>(SlangNative::TypeReflection_GetResourceAccess(m_NativeTypeReflection));
    }    
    System::String^ TypeReflection::Name::get()
    {
        if (!m_NativeTypeReflection) return nullptr;
        return Slang::Cpp::StringUtilities::ToString(SlangNative::TypeReflection_GetName(m_NativeTypeReflection));
    }    
    System::String^ TypeReflection::FullName::get()
    {
        if (!m_NativeTypeReflection) return nullptr;
        // TODO: Add FullName support to DLL exports, for now use Name
        return Slang::Cpp::StringUtilities::ToString(SlangNative::TypeReflection_GetName(m_NativeTypeReflection));
    }
    unsigned int TypeReflection::UserAttributeCount::get()
    {
        if (!m_NativeTypeReflection) return 0;
        return SlangNative::TypeReflection_GetUserAttributeCount(m_NativeTypeReflection);
    }

    Attribute^ TypeReflection::GetUserAttributeByIndex(unsigned int index)
    {
        if (!m_NativeTypeReflection) return nullptr;
        void* attribute = SlangNative::TypeReflection_GetUserAttributeByIndex(m_NativeTypeReflection, index);
        return attribute ? gcnew Attribute(attribute) : nullptr;
    }    Attribute^ TypeReflection::FindAttributeByName(System::String^ name)
    {
        if (!m_NativeTypeReflection) return nullptr;
        const char* nativeName = Slang::Cpp::StringUtilities::FromString(name);
        void* attribute = SlangNative::TypeReflection_FindAttributeByName(m_NativeTypeReflection, nativeName);
        return attribute ? gcnew Attribute(attribute) : nullptr;
    }    Attribute^ TypeReflection::FindUserAttributeByName(System::String^ name)
    {
        if (!m_NativeTypeReflection) return nullptr;
        const char* nativeName = Slang::Cpp::StringUtilities::FromString(name);
        void* attribute = SlangNative::TypeReflection_FindAttributeByName(m_NativeTypeReflection, nativeName);
        return attribute ? gcnew Attribute(attribute) : nullptr;
    }

    TypeReflection^ TypeReflection::ApplySpecializations(GenericReflection^ genRef)
    {
        // TODO: Add ApplySpecializations support to DLL exports
        // For now, return nullptr as this is an advanced feature
        return nullptr;
    }

    GenericReflection^ TypeReflection::GenericContainer::get()
    {
        // TODO: Add GenericContainer support to DLL exports
        // For now, return nullptr as this is an advanced feature
        return nullptr;
    }

    void* TypeReflection::getNative()
    {
        return m_NativeTypeReflection;
    }
    void* TypeReflection::slangPtr()
    {
        return SlangNative::TypeReflection_GetNative(m_NativeTypeReflection);
    }
}
